### PR TITLE
Add Featured Image, Title and Excerpt to top of the settings sidebar

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
- Moves the Featured Image picker to the top of the sidebar.
    - Uses a new dropdown-based flow for selecting a featured image.
    - Adds the ability to upload a featured image using the native file picker.
- Adds new title and excerpt fields to the top of the sidebar.
- Removes the Featured Image and Excerpt panels.

## Why?
Part of https://github.com/WordPress/gutenberg/issues/39082.

## How?
- The Featured Image changes are made in-place using the existing `PostFeaturedImage` component in `@wordpress/editor`.
    - This component is somewhat extensible using a filter that third parties can hook into. I took care to not remove any props that third parties may be using, kept all CSS class names the same, and checked that the new styles do not affect elements that third parties may be adding here.
    - I added the _Upload_ button using `FormFileUpload`.
    - This component was getting very indented so I refactored it into some smaller components. 
- The title and excerpt fields are a new component called `PostSummary` in `@wordpress/edit-post`.
    - It doesn't make sense I think to put these in `@wordpress/editor` as the design we're using is pretty tailored to the post editor sidebar. 
    - Both fields use `EditableText` as rich text formatting is not supported here.
- I removed the Featured Image and Excerpt components from `@wordpress/edit-post` but not `@wordpress/editor` as third parties may be using them.

## Testing Instructions
1. Create or edit a new post.
2. Add a featured image, title, excerpt, whatever!
3. Run the sample code in https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-featured-image/README.md to check that extending the Featured Image still works. 
4. Log in as a contributor to check that these fields don't appear.

## Screenshots or screencast 
https://user-images.githubusercontent.com/612155/178655097-8f8c2123-3e95-4423-b5a6-45b8760a5670.mp4

### CODE REVIEW

**Comment by noisysocks** [comment]
I haven't updated the unit/e2e tests yet but otherwise this should be ready for a first round of review!

File: packages/editor/src/components/post-featured-image/menu.js
Comment (noisysocks) [inline packages/editor/src/components/post-featured-image/menu.js]: Don't think I ended up using this.

File: packages/editor/src/components/post-featured-image/index.js
Comment (Mamaduka) [inline packages/editor/src/components/post-featured-image/index.js:262]: Thank you. I forgot to include the comment when during my last refactor.

**Comment by noisysocks** [comment]
Reviewing the feedback from when we last implemented something similar to this in https://github.com/WordPress/gutenberg/pull/39973 (which I totally forgot about lol):

- https://github.com/WordPress/gutenberg/pull/39973#issuecomment-1088920081
- https://github.com/WordPress/gutenberg/pull/40128#issuecomment-1091114713

These were the concerns that I can see were raised:

1. Problems with the loading spinner in the _Featured Image_ control. @Mamaduka has since improved this.

1. Lack of post type support checks, meaning that _Featured Image_, _Title_, and _Excerpt_ would appear for users which did not have permission to edit them. This PR handles this correctly.

2. Not clear that _Title_ and _Excerpt_ are editable fields (https://github.com/WordPress/gutenberg/issues/40124). This PR still has that issue.
   
   @javierarce, perhaps we could add a slight border to the title and excerpt fields? Or return to the more modal UI that @shaunandrews suggested.

   Intro

3. Too much prominence given to _Excerpt_. This PR still has that issue. 

   I think implementing a dropdown menu like the one proposed by @javierarce in https://github.com/WordPress/gutenberg/pull/42030#issuecomment-1171646574 would work well at addressing this. Each of the fields in _Summary_ can be configurable and we can default to having _Excerpt_ disabled.

   https://user-images.githubusercontent.com/4933/176771391-8a450f38-068b-451d-af19-9430a7f51a20.mov

   It's not clear how this would fit in with https://github.com/WordPress/gutenberg/pull/42175 though.

4. Not enough time to test this in the plugin before WordPress 6.0. We're still a way away from WordPress 6.1 so I think this is okay. We can comfortably address the above issues in follow-up PRs.

**Review by Mamaduka — COMMENTED** [review]
Thanks for working on this, @noisysocks!

✅ Tested that all new fields work.
✅ The featured image filters still apply. While filters still work, do we maintain the same CSS classes so we don't break styling for overrides? 
✅  I can't see Featured Image as a contributor.

## Notes
* I noticed an empty gap with borders when the post type doesn't support any of these fields (see the screenshot).
* What do you think about using `object-fit: cover` for the featured image?

> Uses a new dropdown-based flow for selecting a featured image.

It is a big difference from the #39973. While I don't mind this new flow, considering the recent discussion in #42175, do you think we should leave the old image upload UX?


## Screenshot
CleanShot 2022-07-13 at 16 58 58

**Comment by ndiego** [comment]
Very excited for this, will do some testing later this week!

**Comment by noisysocks** [comment]
> * I noticed an empty gap with borders when the post type doesn't support any of these fields (see the screenshot).

Oops! Fixed in [67d9a55](https://github.com/WordPress/gutenberg/pull/42352/commits/67d9a555d12c8eccda07bdbe0ea1b0dd0ea00f2e).

> * What do you think about using `object-fit: cover` for the featured image?
> 
> It is a big difference from the #39973. While I don't mind this new flow, considering the recent discussion in #42175, do you think we should leave the old image upload UX?

I'll defer to @javierarce on these two questions 😀

Personally I really like the feel of this featured image flow. It lets us add an _Upload_ button, which is very convenient, without forcing the user to make decisions before they have signalled an intent to change the featured image.

File: packages/editor/src/components/post-excerpt/check.js
Comment (Mamaduka) [inline packages/editor/src/components/post-excerpt/check.js]: What do you think about using `usePostTypeSupportCheck( 'excerpt' )` directly? One less public API to maintainn for us.

P.S. I don't think I've seen hooks used in ternaries before 😄

Comment (noisysocks) [reply packages/editor/src/components/post-excerpt/check.js]: Yeah, I added it for consistency, but I suppose they're both useless. I'll remove `usePostExcerptCheck` and deprecate `PostExcerptCheck`.

Comment (Mamaduka) [reply packages/editor/src/components/post-excerpt/check.js]: I think we can keep `PostExcerptCheck` and just use `usePostTypeSupportCheck` internally. What do you think?

**Comment by noisysocks** [comment]
> I haven't updated the unit/e2e tests yet but otherwise this should be ready for a first round of review!

I've updated the tests! 🤞

**Comment by noisysocks** [comment]
@javierarce @Mamaduka: How are we feeling about this? I'd like to merge it this week and then work on points 2 and 3 in https://github.com/WordPress/gutenberg/pull/42352#issuecomment-1182813156 as follow-up PRs once we've figured out https://github.com/WordPress/gutenberg/pull/42175.

**Review by Mamaduka — APPROVED** [review]
@noisysocks, I think PR is in good shape to merge.

My only minor concern is using popup for upload actions (https://github.com/WordPress/gutenberg/pull/42352#pullrequestreview-1037332448), but I would defer to @javierarce on that matter.

**Comment by paaljoachim** [comment]
Regarding excerpt. 
Should it contain styling options such as bold etc? So that we can style the excerpt. (As if it was a Paragraph block.)

**Comment by noisysocks** [comment]
Excerpt is similar to title in that it's up to the theme whether HTML is supported or if it's escaped before being rendered. It would be a good enhancement to make but requires some careful thought about how themes can indicate that they support HTML excerpts (maybe using `add_theme_supports`). In other words, out of scope for this PR I think! 😀

**Comment by javierarce** [comment]
This works great, great job!

That said, there are a couple of things we need to address before merging it: 

First of all, the accessibility of the text fields. This @carolinan  mentioned [here](https://github.com/WordPress/gutenberg/issues/40124) and we should address it. In the design file, I proposed using an edit icon that appears on hover, but that might not be enough. I'll try to think of a solution for that.

The other thing is the placement of the popover. Right now we show it besides the sidebar (and vertically centered when there's an image). 

Screenshot 2022-07-18 at 16 35 54
Screenshot 2022-07-18 at 16 37 43

Since we always open popovers inside the sidebar, we should do the same thing here and place the popover under the link or the edit icon.

**Comment by paaljoachim** [comment]
Good work Robert @noisysocks !

Hmm my initial reaction is some hesitancy. I brought a screenshot into Affinity Photo and reordered fields, and came to the conclusion that the current UI does work well. Looking at it side by side is interesting though.

Here is a mockup and the current exploration.
Current and new exploration

I would though when hovering over the featured image add a popup/tooltip saying "Featured Image"

As have been mentioned above. Clicking into the title and excerpt it would be helpful to include a thin outline giving the signal that it is editable.

### The side popup.

Screenshot 2022-07-18 at 18 17 18

I went ahead and compared it with clicking the Color: Text, Background and Link.
Clicking any of these the popup box would align along the top of each.
Like so:
Screenshot 2022-07-18 at 18 19 35

It would be nice to also align the Featured image popup along the top edge/line of Featured image area.

**Comment by noisysocks** [comment]
> Since we always open popovers inside the sidebar, we should do the same thing here and place the popover under the link or the edit icon.

👍 I've changed the popover to appear below the featured image button.

Screen Shot 2022-07-19 at 13 02 47
Screen Shot 2022-07-19 at 13 02 33

> I would though when hovering over the featured image add a popup/tooltip saying "Featured Image"

I agreed with this but then after implementing it I changed my mind. I don't think a label adds any value. There is already a label that appears in the popover when you click on the image, there is already a pencil icon which appears on hover, and there is already an `aria-label` which is read aloud for users using assistive technologies.

Here's what it looked like:

https://user-images.githubusercontent.com/612155/179655581-5c445e49-01d5-4fad-9612-d3c7923af909.mp4

> First of all, the accessibility of the text fields. This @carolinan mentioned https://github.com/WordPress/gutenberg/issues/40124 and we should address it. In the design file, I proposed using an edit icon that appears on hover, but that might not be enough. I'll try to think of a solution for that.
> As have been mentioned above. Clicking into the title and excerpt it would be helpful to include a thin outline giving the signal that it is editable.

Would you mind if we did this in a follow-up? It's something that requires design feedback which I think would be easier to obtain in an issue or a PR that is solely focused on addressing this.

**Comment by paaljoachim** [comment]
I was going to say that it would be nice to open popups on the left side of the sidebar (similar to Color options), but then I began clicking Visibility, Publish, URL and Template and noticed they all opened below each option. Clicking Set featured image and having the popup show up in the same way keeps a consistent way of handling the popups. 

One thing that came up. Should there be a placeholder outline around the "Set featured image" text? (As I think more about it...there is probably no need for it.)

I will add the above comment even though it really does not add to the discussion. Good work Robert!

**Comment by jameskoster** [comment]
One small nitpick on the design:

Screenshot 2022-07-21 at 16 59 36

The pencil icon almost tricks me into believing I have to click it to edit, despite being able to click the entire image. I wonder if there's a hover effect we could use that better communicates that the entire area is clickable, kind of like when the featured image is not set?

Apologies if this already came up :)

**Comment by jameskoster** [comment]
And one small issue I noticed with the excerpt – if you edit via the Inspector, any formatting you had is stripped:


https://user-images.githubusercontent.com/846565/180260891-6a584b40-1c69-4714-b794-551876fec834.mp4

This inevitably raises the question – should there be formatting options for the excerpt in the Inspector?

**Comment by critterverse** [comment]
Hey all, was wondering if we could give this a try as a part of the pre-publish panel instead? It feels like we're running into some of the same issues as the [previous exploration](https://github.com/WordPress/gutenberg/pull/39973) that was reverted. I left a comment on https://github.com/WordPress/gutenberg/issues/39082#issuecomment-1194089371 as well, let me know what you think.

**Comment by noisysocks** [comment]
Yep, let's close this out since there's consensus on trying something else. See https://github.com/WordPress/gutenberg/issues/39082#issuecomment-1195279306.

**Pull request closed without merging**
